diff --git a/zaoai/src/app.rs b/zaoai/src/app.rs
index 753e99e..1ab8cd6 100644
--- a/zaoai/src/app.rs
+++ b/zaoai/src/app.rs
@@ -258,9 +258,10 @@ impl eframe::App for ZaoaiApp {
                             self.training_thread.rx_validation_payload = None;
                         }
 
-                        if received_any_training || received_any_validation {
-                            ctx.request_repaint();
-                        }
+                        // if received_any_training || received_any_validation {
+                        // Need to repaint each frame for now. Otherwise crash due to channel disconnect on complete
+                        ctx.request_repaint();
+                        // }
 
                         let training_in_progress = self.training_thread.training_in_progress();
                         if !training_in_progress && received_any_training {
diff --git a/zaoai/src/app_windows.rs b/zaoai/src/app_windows.rs
index e246ef2..ba2436f 100644
--- a/zaoai/src/app_windows.rs
+++ b/zaoai/src/app_windows.rs
@@ -1,14 +1,10 @@
 use std::{ ops::RangeInclusive, path::PathBuf, sync::{Arc, Mutex, mpsc::{self, Receiver, Sender}}, thread::JoinHandle};
 
 use crate::{
-    app::{AppState, MenuWindowData},
-    egui_ext::{add_slider_sized, Interval},
-    mnist::get_mnist,
-    zneural_network::{
+    app::{AppState, MenuWindowData}, egui_ext::{Interval, add_slider_sized}, mnist::get_mnist, zneural_network::{
         activation::ActivationFunctionType, cost::CostFunction, datapoint::{
-            create_2x2_test_datapoints,  DataPoint, TrainingData, TrainingDataset,
-            VirtualTrainingDataset,
-        }, is_correct::ConfusionEvaluator, layer::{ BiasInit, WeightInit}, neuralnetwork::{GraphStructure, NeuralNetwork}, thread::{TrainingThreadController, TrainingThreadPayload}, training::{test_nn, FloatDecay, TestResults, TrainingSession, TrainingState}
+            DataPoint, TrainingData, TrainingDataset, VirtualTrainingDataset, create_2x2_test_datapoints,
+        }, is_correct::ConfusionEvaluator, layer::{ BiasInit, WeightInit}, neuralnetwork::{GraphStructure, NeuralNetwork, NeuralNetworkPingPong}, thread::{TrainingThreadController, TrainingThreadPayload}, training::{FloatDecay, TestResults, TrainingSession, TrainingState, test_nn}
     },
 };
 use eframe::egui::{self, Align, Button, Color32, InnerResponse, Layout, Sense, Slider, SliderClamping};
@@ -449,7 +445,8 @@ impl<'a> DrawableWindow<'a> for WindowAi {
                                     let training_data_clone = training_data.clone();
                                     self.test_nn_handle = Some(std::thread::spawn(move || {
                                         log::trace!("Test Thread test_nn spawned!");
-                                        match test_nn(&mut ai_clone, &training_data_clone.test_split(), is_correct_fn, maybe_tx, Some(rx_abort))
+                                        let pingpong = &mut NeuralNetworkPingPong::new(ai_clone.max_layer_nodes());
+                                        match test_nn(&mut ai_clone, &training_data_clone.test_split(), is_correct_fn, maybe_tx, Some(rx_abort), pingpong)
                                         {
                                             Ok(r) => {
                                                 log::trace!("Test Thread test_nn complete!");
diff --git a/zaoai/src/zneural_network/activation.rs b/zaoai/src/zneural_network/activation.rs
index f5c800d..ed549f7 100644
--- a/zaoai/src/zneural_network/activation.rs
+++ b/zaoai/src/zneural_network/activation.rs
@@ -17,10 +17,7 @@ pub enum ActivationFunctionType {
 
 impl ActivationFunctionType {
     #[cfg(feature = "simd")]
-    pub fn apply_softmax(layer_values: &[f32]) -> Vec<f32> {
-        let len = layer_values.len();
-        let mut output = vec![0.0f32; len];
-
+    pub fn apply_softmax(layer_values: &mut [f32]) {
         let max_val = layer_values
             .iter()
             .copied()
@@ -33,6 +30,7 @@ impl ActivationFunctionType {
         for chunk in chunks {
             use wide::f32x8;
 
+            // TODO: as_mut_chunk
             let v = f32x8::from(chunk);
             let e = (v - f32x8::splat(max_val)).exp();
             let temp = e.to_array();
@@ -48,11 +46,9 @@ impl ActivationFunctionType {
         let sum_f32 = sum as f32;
 
         // Normalize
-        for (i, &val) in layer_values.iter().enumerate() {
-            output[i] = (val - max_val).exp() / sum_f32;
+        for val in layer_values.iter_mut() {
+            *val = (*val - max_val).exp() / sum_f32;
         }
-
-        output
     }
     #[cfg(not(feature = "simd"))]
     pub fn apply_softmax(layer_values: &[f32]) -> Vec<f32> {
diff --git a/zaoai/src/zneural_network/cost.rs b/zaoai/src/zneural_network/cost.rs
index ea1bab3..86af4a1 100644
--- a/zaoai/src/zneural_network/cost.rs
+++ b/zaoai/src/zneural_network/cost.rs
@@ -23,13 +23,14 @@ impl CostFunction {
     }
     pub fn call_d(&self, predicted: &[f32], expected: &[f32]) -> f32 {
         match self {
-            CostFunction::Mse => mse(predicted, expected),
+            CostFunction::Mse => mse_d(predicted, expected),
             CostFunction::CrossEntropyBinary => cross_entropy_loss_binary_d(predicted, expected),
             CostFunction::CrossEntropyMulticlass => {
-                cross_entropy_loss_multiclass(predicted, expected)
+                cross_entropy_loss_multiclass_d(predicted, expected)
             }
         }
     }
+    #[cfg(feature = "simd")]
     pub fn call_simd(&self, predicted: f32x8, expected: f32x8) -> f32x8 {
         match self {
             CostFunction::Mse => mse_simd(predicted, expected),
@@ -39,6 +40,7 @@ impl CostFunction {
             }
         }
     }
+    #[cfg(feature = "simd")]
     pub fn call_simd_d(&self, predicted: f32x8, expected: f32x8) -> f32x8 {
         match self {
             CostFunction::Mse => mse_d_simd(predicted, expected),
diff --git a/zaoai/src/zneural_network/layer.rs b/zaoai/src/zneural_network/layer.rs
index 14a206e..197fc77 100644
--- a/zaoai/src/zneural_network/layer.rs
+++ b/zaoai/src/zneural_network/layer.rs
@@ -10,9 +10,10 @@ use serde::{Deserialize, Serialize};
 use strum_macros::Display;
 use wide::f32x8;
 
-use crate::zneural_network::activation::ActivationFunctionType;
-#[cfg(feature = "simd")]
 use crate::zneural_network::cost::CostFunction;
+use crate::zneural_network::{
+    activation::ActivationFunctionType, neuralnetwork::NeuralNetworkPingPong,
+};
 
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 #[derive(Clone, Debug, bincode::Encode, bincode::Decode)]
@@ -317,34 +318,29 @@ impl Layer {
     }
 
     #[cfg(not(feature = "simd"))]
-    pub fn apply_activation(weighted_inputs: &[f32], t: ActivationFunctionType) -> Vec<f32> {
-        weighted_inputs.iter().map(|&x| t.activate(x)).collect()
+    pub fn apply_activation(weighted_inputs: &mut [f32], t: ActivationFunctionType) {
+        weighted_inputs.iter_mut().for_each(|x| *x = t.activate(*x));
     }
     #[cfg(feature = "simd")]
-    pub fn apply_activation(input: &[f32], t: ActivationFunctionType) -> Vec<f32> {
+    pub fn apply_activation(input: &mut [f32], t: ActivationFunctionType) {
         const CHUNK_SIZE: usize = 8;
 
-        let len = input.len();
-        let mut result = vec![0f32; len];
+        let mut chunks = input.chunks_exact_mut(CHUNK_SIZE);
+        // let mut chunks = input.as_chunks_mut::<CHUNK_SIZE>().0;
 
-        let chunks = input.chunks_exact(CHUNK_SIZE);
-        let remainder = chunks.remainder();
+        for chunk in chunks.by_ref() {
+            let arr: [f32; CHUNK_SIZE] = (&*chunk).try_into().unwrap();
+            let input_vec = f32x8::from(arr);
 
-        for (i, chunk) in chunks.enumerate() {
-            let input_vec = f32x8::from(chunk);
             let activated_vec = t.activate_simd(input_vec);
             let out: [f32; CHUNK_SIZE] = activated_vec.into();
 
-            let start = i * CHUNK_SIZE;
-            result[start..start + CHUNK_SIZE].copy_from_slice(&out);
+            chunk.copy_from_slice(&out);
         }
 
-        let rem_start = len - remainder.len();
-        for (i, &x) in remainder.iter().enumerate() {
-            result[rem_start + i] = t.activate(x);
+        for x in chunks.into_remainder() {
+            *x = t.activate(*x);
         }
-
-        result
     }
 
     #[cfg(not(feature = "simd"))]
@@ -401,66 +397,61 @@ impl Layer {
         }
     }
 
-    pub fn calculate_outputs(&self, inputs: &[f32]) -> Vec<f32> {
-        let mut weighted_inputs = vec![0.0; self.num_out_nodes];
-        self.compute_weighted_inputs_scalar(inputs, &mut weighted_inputs);
-        Self::apply_activation(&weighted_inputs, self.activation_type)
+    #[cfg(not(feature = "simd"))]
+    pub fn calculate_outputs(&self, inputs: &[f32], outputs: &mut [f32]) {
+        self.compute_weighted_inputs_scalar(inputs, outputs);
+        Self::apply_activation(outputs, self.activation_type);
     }
     #[cfg(feature = "simd")]
-    pub fn calculate_outputs_simd(&self, inputs: &[f32]) -> Vec<f32> {
-        let mut weighted_inputs = vec![0.0; self.num_out_nodes];
-        self.compute_weighted_inputs_simd(inputs, &mut weighted_inputs);
-        Self::apply_activation(&weighted_inputs, self.activation_type)
+    pub fn calculate_outputs_simd(&self, inputs: &[f32], outputs: &mut [f32]) {
+        self.compute_weighted_inputs_simd(inputs, outputs);
+        Self::apply_activation(outputs, self.activation_type);
     }
     #[allow(dead_code)]
     #[cfg(feature = "simd")]
-    pub fn calculate_outputs_simd_rayon(&self, inputs: &[f32]) -> Vec<f32> {
-        let mut weighted_inputs = vec![0.0; self.num_out_nodes];
-        self.compute_weighted_inputs_simd_rayon(inputs, &mut weighted_inputs);
-        Self::apply_activation(&weighted_inputs, self.activation_type)
+    pub fn calculate_outputs_simd_rayon(&self, inputs: &[f32], outputs: &mut [f32]) {
+        self.compute_weighted_inputs_simd_rayon(inputs, outputs);
+        Self::apply_activation(outputs, self.activation_type);
     }
 
     pub fn calculate_outputs_learn(
         &mut self,
         inputs: &[f32],
+        outputs: &mut [f32],
         learn_data: &mut LayerLearnData,
-    ) -> Vec<f32> {
-        learn_data.inputs.clone_from_slice(inputs);
-
-        let mut weighted_inputs = vec![0.0; self.num_out_nodes];
-        self.compute_weighted_inputs_scalar(inputs, &mut weighted_inputs);
+    ) {
+        learn_data.inputs.clear();
+        learn_data.inputs.extend_from_slice(inputs);
 
-        self.fill_learn_data(learn_data, &weighted_inputs);
-        learn_data.activation_values.clone()
+        self.compute_weighted_inputs_scalar(inputs, outputs);
+        self.fill_learn_data(learn_data, &outputs);
     }
     #[cfg(feature = "simd")]
     pub fn calculate_outputs_learn_simd(
         &mut self,
         inputs: &[f32],
+        outputs: &mut [f32],
         learn_data: &mut LayerLearnData,
-    ) -> Vec<f32> {
-        learn_data.inputs.clone_from_slice(inputs);
-
-        let mut weighted_inputs = vec![0.0; self.num_out_nodes];
-        self.compute_weighted_inputs_simd(inputs, &mut weighted_inputs);
+    ) {
+        learn_data.inputs.clear();
+        learn_data.inputs.extend_from_slice(inputs);
 
-        self.fill_learn_data(learn_data, &weighted_inputs);
-        learn_data.activation_values.clone()
+        self.compute_weighted_inputs_simd(inputs, outputs);
+        self.fill_learn_data(learn_data, &outputs);
     }
     #[allow(dead_code)]
     #[cfg(feature = "simd")]
     pub fn calculate_outputs_learn_simd_rayon(
         &mut self,
         inputs: &[f32],
+        outputs: &mut [f32],
         learn_data: &mut LayerLearnData,
-    ) -> Vec<f32> {
-        learn_data.inputs.clone_from_slice(inputs);
-
-        let mut weighted_inputs = vec![0.0; self.num_out_nodes];
-        self.compute_weighted_inputs_simd_rayon(inputs, &mut weighted_inputs);
+    ) {
+        learn_data.inputs.clear();
+        learn_data.inputs.extend_from_slice(inputs);
 
-        self.fill_learn_data(learn_data, &weighted_inputs);
-        learn_data.activation_values.clone()
+        self.compute_weighted_inputs_simd_rayon(inputs, outputs);
+        self.fill_learn_data(learn_data, &outputs);
     }
 
     pub fn apply_cost_gradient(&mut self, learn_rate: f32) {
@@ -609,7 +600,10 @@ impl Layer {
         cost_fn: CostFunction,
     ) {
         for i in 0..learn_data.node_values.len() {
-            let dcost = cost_fn.call_d(learn_data.activation_values[i], expected_outputs[i]);
+            let dcost = cost_fn.call_d(
+                &vec![learn_data.activation_values[i]],
+                &vec![expected_outputs[i]],
+            );
             let dactivation = self
                 .activation_type
                 .activate_derivative(learn_data.weighted_inputs[i]);
diff --git a/zaoai/src/zneural_network/neuralnetwork.rs b/zaoai/src/zneural_network/neuralnetwork.rs
index 570f601..8a4b85d 100644
--- a/zaoai/src/zneural_network/neuralnetwork.rs
+++ b/zaoai/src/zneural_network/neuralnetwork.rs
@@ -97,6 +97,20 @@ impl GraphStructure {
     }
 }
 
+pub struct NeuralNetworkPingPong {
+    pub current: Vec<f32>, // in
+    pub next: Vec<f32>,    // out
+}
+
+impl NeuralNetworkPingPong {
+    pub fn new(max_layer_nodes: usize) -> Self {
+        Self {
+            current: Vec::with_capacity(max_layer_nodes),
+            next: Vec::with_capacity(max_layer_nodes),
+        }
+    }
+}
+
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 #[derive(Clone, bincode::Encode, bincode::Decode)]
 pub struct NeuralNetwork {
@@ -167,6 +181,21 @@ impl NeuralNetwork {
         }
     }
 
+    pub fn max_layer_nodes(&self) -> usize {
+        let hidden_layers_max = *self
+            .graph_structure
+            .hidden_layers
+            .iter()
+            .max()
+            .unwrap_or(&0);
+        let max = self
+            .graph_structure
+            .input_nodes
+            .max(self.graph_structure.output_nodes)
+            .max(hidden_layers_max);
+        max
+    }
+
     fn apply_dropout(inputs: &mut [f32], mask: &mut Vec<f32>, dropout_prob: f32) {
         let keep_prob = 1.0 - dropout_prob;
         let mut rng = rand::thread_rng();
@@ -188,6 +217,7 @@ impl NeuralNetwork {
         learn_rate: f32,
         batch_data_cost: &mut f32,
         batch_data_loss: &mut f32,
+        pingpong: &mut NeuralNetworkPingPong,
     ) -> Vec<Vec<f32>> {
         if batch_data.len() <= 0 {
             panic!("DataPoints length was 0");
@@ -197,17 +227,16 @@ impl NeuralNetwork {
         let mut last_loss = 0.0; // last batches cost
         let mut batch_data_outputs = Vec::with_capacity(batch_data.len());
         for (i, datapoint) in batch_data.iter().enumerate() {
-            let datapoint_outputs = self.learn_calculate_outputs(datapoint);
+            self.learn_calculate_outputs(datapoint, pingpong);
             // let loss =
             //     cross_entropy_loss_multiclass(&datapoint_outputs, &datapoint.expected_outputs);
-            let cost = self.cost_function(&datapoint_outputs, &datapoint.expected_outputs);
+            let cost = self.cost_function(&pingpong.next, &datapoint.expected_outputs);
 
             total_cost += cost;
             if i == batch_data.len() - 1 {
                 last_loss = cost;
             }
-
-            batch_data_outputs.push(datapoint_outputs);
+            batch_data_outputs.push(pingpong.next.clone());
         }
         // Adjust weights & biases
         self.apply_all_cost_gradients(learn_rate / (batch_data.len() as f32));
@@ -229,6 +258,7 @@ impl NeuralNetwork {
         learn_rate: f32,
         is_correct_fn: ConfusionEvaluator,
         mut epoch_metadata: Option<&mut AIResultMetadata>,
+        pingpong: &mut NeuralNetworkPingPong,
     ) {
         assert!(!training_data.is_empty());
         assert_eq!(
@@ -256,8 +286,13 @@ impl NeuralNetwork {
 
                 let mut batch_data_cost = 0.0;
                 let mut batch_data_loss = 0.0;
-                let batch_data_outputs =
-                    self.learn_batch(data, learn_rate, &mut batch_data_cost, &mut batch_data_loss);
+                let batch_data_outputs = self.learn_batch(
+                    data,
+                    learn_rate,
+                    &mut batch_data_cost,
+                    &mut batch_data_loss,
+                    pingpong,
+                );
 
                 if let Some(metadata) = epoch_metadata.as_mut() {
                     let mut new_metadata = AIResultMetadata::new(
@@ -341,10 +376,13 @@ impl NeuralNetwork {
         assert!(training_data.len() > 0);
         assert!(batch_size > 0);
 
+        let pingpong = &mut NeuralNetworkPingPong::new(self.max_layer_nodes());
+
         for e in 0..num_epochs {
             let mut test_nn_and_send_payload =
                 |tx: &Sender<TrainingThreadPayload>, data: &[DataPoint], payload_index: usize| {
-                    if let Some(test_results) = test_nn(self, data, is_correct_fn, None, None).ok()
+                    if let Some(test_results) =
+                        test_nn(self, data, is_correct_fn, None, None, pingpong).ok()
                     {
                         let mut result_metadata = AIResultMetadata::from_accuracy(
                             test_results.accuracy.unwrap_or_default() as f64,
@@ -396,6 +434,7 @@ impl NeuralNetwork {
                 maybe_decayed_learn_rate,
                 is_correct_fn,
                 Some(&mut metadata),
+                pingpong,
             );
 
             if tx_training_metadata.is_some() {
@@ -421,34 +460,39 @@ impl NeuralNetwork {
         log::info!("Training...Complete! [@{} Epochs]", num_epochs);
     }
 
-    pub fn learn_calculate_outputs(&mut self, datapoint: &DataPoint) -> Vec<f32> {
-        let outputs = self.forward(&datapoint.inputs);
+    pub fn learn_calculate_outputs(
+        &mut self,
+        datapoint: &DataPoint,
+        pingpong: &mut NeuralNetworkPingPong,
+    ) {
+        self.forward(&datapoint.inputs, pingpong);
         self.backpropagation(datapoint);
-        outputs
+        // Now results should be in pingpong.next
     }
 
-    fn forward(&mut self, inputs: &[f32]) -> Vec<f32> {
-        let mut current_inputs = inputs.to_vec();
+    fn forward(&mut self, inputs: &[f32], pingpong: &mut NeuralNetworkPingPong) {
+        pingpong.current.clear();
+        pingpong.current.extend_from_slice(inputs);
 
         for (i, layer) in self.layers.iter_mut().enumerate() {
-            let learn_data = &mut self.layer_learn_data[i];
-            learn_data.inputs.reserve(current_inputs.len());
+            pingpong.next.resize(layer.num_out_nodes, 0.0);
 
-            current_inputs = if cfg!(feature = "simd") {
-                layer.calculate_outputs_learn_simd(&mut current_inputs, learn_data)
-            } else {
-                layer.calculate_outputs_learn(&mut current_inputs, learn_data)
-            };
+            let learn_data = &mut self.layer_learn_data[i];
+            #[cfg(feature = "simd")]
+            layer.calculate_outputs_learn_simd(&pingpong.current, &mut pingpong.next, learn_data);
+            #[cfg(not(feature = "simd"))]
+            layer.calculate_outputs_learn(&pingpong.current, &mut pingpong.next, learn_data);
 
             if let Some(prob) = layer.dropout_prob {
                 let mask = learn_data
                     .dropout_mask
-                    .get_or_insert_with(|| vec![0.0; current_inputs.len()]);
-                Self::apply_dropout(&mut current_inputs, mask, prob);
+                    .get_or_insert_with(|| vec![0.0; pingpong.next.len()]);
+                Self::apply_dropout(&mut pingpong.next, mask, prob);
             }
+            std::mem::swap(&mut pingpong.current, &mut pingpong.next);
         }
-
-        current_inputs
+        // For simplicity, keep input at current, output at next
+        std::mem::swap(&mut pingpong.current, &mut pingpong.next);
     }
 
     fn backpropagation(&mut self, datapoint: &DataPoint) {
@@ -506,33 +550,59 @@ impl NeuralNetwork {
         }
     }
 
-    pub fn calculate_outputs(&self, inputs: &[f32]) -> Vec<f32> {
-        let mut current_inputs = inputs.to_vec();
-        for (_, layer) in self.layers.iter().enumerate() {
+    // pub fn calculate_outputs(&self, inputs: &[f32]) -> Vec<f32> {
+    //     let mut current_inputs = inputs.to_vec();
+    //     for (_, layer) in self.layers.iter().enumerate() {
+    //         #[cfg(feature = "simd")]
+    //         {
+    //             current_inputs = layer.calculate_outputs_simd(&current_inputs);
+    //         }
+    //         #[cfg(not(feature = "simd"))]
+    //         {
+    //             current_inputs = layer.calculate_outputs(&current_inputs);
+    //         }
+    //     }
+
+    //     if self.is_softmax_output {
+    //         ActivationFunctionType::apply_softmax(&mut current_inputs);
+    //     }
+    //     current_inputs
+    // }
+    pub fn calculate_outputs(&self, inputs: &[f32], pingpong: &mut NeuralNetworkPingPong) {
+        pingpong.current.clear();
+        pingpong.current.extend_from_slice(inputs);
+
+        for layer in &self.layers {
+            pingpong.next.resize(layer.num_out_nodes, 0.0);
+
             #[cfg(feature = "simd")]
-            {
-                current_inputs = layer.calculate_outputs_simd(&current_inputs);
-            }
+            layer.calculate_outputs_simd(&pingpong.current, &mut pingpong.next);
             #[cfg(not(feature = "simd"))]
-            {
-                current_inputs = layer.calculate_outputs(&current_inputs);
-            }
+            layer.calculate_outputs(&pingpong.current, &mut pingpong.next);
+
+            std::mem::swap(&mut pingpong.current, &mut pingpong.next);
         }
 
+        // Simplicity, keep input at current, output at next
+        std::mem::swap(&mut pingpong.current, &mut pingpong.next);
+
         if self.is_softmax_output {
-            current_inputs = ActivationFunctionType::apply_softmax(&current_inputs)
+            ActivationFunctionType::apply_softmax(&mut pingpong.next);
         }
-        current_inputs
     }
 
     fn cost_function(&self, predicted: &[f32], expected: &[f32]) -> f32 {
         self.cost_fn.call(predicted, expected)
     }
     #[cfg(not(feature = "simd"))]
-    fn calculate_cost_datapoint(&self, datapoint: &DataPoint) -> f32 {
+    fn calculate_cost_datapoint(
+        &self,
+        datapoint: &DataPoint,
+        pingpong: &mut NeuralNetworkPingPong,
+    ) -> f32 {
         // Prediction cost
-        let outputs = self.calculate_outputs(&datapoint.inputs);
-        let cost = self.cost_function(&outputs, &datapoint.expected_outputs);
+        self.calculate_outputs(&datapoint.inputs, pingpong);
+        let cost = self.cost_function(&pingpong.next, &datapoint.expected_outputs);
 
         // L2 regularization penalty (sum of squared weights)
         let l2_penalty: f32 = self
@@ -546,41 +616,41 @@ impl NeuralNetwork {
         const LAMBDA: f32 = 0.001;
         cost + LAMBDA * l2_penalty
     }
-    pub fn calculate_costs(&self, data: &[DataPoint]) -> f32 {
+    pub fn calculate_costs(&self, data: &[DataPoint], pingpong: &mut NeuralNetworkPingPong) -> f32 {
         assert!(!data.is_empty(), "Input data was empty");
 
         #[cfg(feature = "simd")]
         {
-            self.calculate_cost_simd(data)
+            self.calculate_cost_simd(data, pingpong)
         }
         #[cfg(not(feature = "simd"))]
         {
-            self.calculate_cost(data)
+            self.calculate_cost(data, pingpong)
         }
     }
     #[cfg(not(feature = "simd"))]
-    fn calculate_cost(&self, data: &[DataPoint]) -> f32 {
+    fn calculate_cost(&self, data: &[DataPoint], pingpong: &mut NeuralNetworkPingPong) -> f32 {
         let total: f32 = data
             .iter()
-            .map(|dp| self.calculate_cost_datapoint(dp))
+            .map(|dp| self.calculate_cost_datapoint(dp, pingpong))
             .sum();
 
         total / (data.len() as f32)
     }
     #[cfg(feature = "simd")]
-    fn calculate_cost_simd(&self, data: &[DataPoint]) -> f32 {
+    fn calculate_cost_simd(&self, data: &[DataPoint], pingpong: &mut NeuralNetworkPingPong) -> f32 {
         let output_layer = self.layers.last().unwrap();
         let num_outputs = output_layer.num_out_nodes;
 
         let total_cost: f32 = data
             .iter()
             .map(|datapoint| {
-                let output = self.calculate_outputs(&datapoint.inputs);
+                self.calculate_outputs(&datapoint.inputs, pingpong);
                 let mut sum = f32x8::splat(0.0);
                 let mut i = 0;
 
                 while i + 8 <= num_outputs {
-                    let pred = f32x8::from(&output[i..i + 8]);
+                    let pred = f32x8::from(&pingpong.next[i..i + 8]);
                     let expected = f32x8::from(&datapoint.expected_outputs[i..i + 8]);
                     sum += self.cost_fn.call_simd(pred, expected);
                     i += 8;
@@ -590,7 +660,7 @@ impl NeuralNetwork {
                 if i < num_outputs {
                     cost += self
                         .cost_fn
-                        .call(&output[i..], &datapoint.expected_outputs[i..]);
+                        .call(&pingpong.next[i..], &datapoint.expected_outputs[i..]);
                 }
                 cost
             })
diff --git a/zaoai/src/zneural_network/training.rs b/zaoai/src/zneural_network/training.rs
index 9208528..3f2d325 100644
--- a/zaoai/src/zneural_network/training.rs
+++ b/zaoai/src/zneural_network/training.rs
@@ -13,7 +13,7 @@ use strum_macros::Display;
 use crate::zneural_network::{
     datapoint::{DataPoint, TrainingData},
     is_correct::ConfusionEvaluator,
-    neuralnetwork::{NNOutputs, NeuralNetwork},
+    neuralnetwork::{NNOutputs, NeuralNetwork, NeuralNetworkPingPong},
     thread::TrainingThreadPayload,
 };
 
@@ -336,6 +336,7 @@ pub fn test_nn<'a>(
     is_correct_fn: ConfusionEvaluator,
     tx_test_metadata: Option<Sender<TrainingThreadPayload>>,
     tx_abort: Option<Receiver<()>>,
+    pingpong: &mut NeuralNetworkPingPong,
 ) -> Result<&'a TestResults, anyhow::Error> {
     if test_data.len() >= 1
         && test_data.first().unwrap().inputs.len() == nn.graph_structure.input_nodes
@@ -348,14 +349,12 @@ pub fn test_nn<'a>(
         let mut results = Vec::with_capacity(test_data.len());
         for i in 0..test_data.len() {
             let datapoint = &test_data[i];
-            let outputs = nn.calculate_outputs(&datapoint.inputs[..]);
-
+            let cost = nn.calculate_costs(std::slice::from_ref(&test_data[i]), pingpong);
             if let Some(tx_test_metadata) = &tx_test_metadata {
-                let cost = nn.calculate_costs(std::slice::from_ref(&test_data[i]));
                 let mut metadata_point =
                     AIResultMetadata::new(DatasetUsage::Test, cost as f64, cost as f64, 0.0);
 
-                let confusion = is_correct_fn.evaluate(&outputs, &datapoint.expected_outputs);
+                let confusion = is_correct_fn.evaluate(&pingpong.next, &datapoint.expected_outputs);
                 match confusion {
                     crate::zneural_network::is_correct::ConfusionCategory::TruePositive => {
                         metadata_point.true_positives += 1
@@ -389,10 +388,11 @@ pub fn test_nn<'a>(
                 }
             }
 
-            results.push((test_data[i].clone(), outputs));
+            results.push((test_data[i].clone(), pingpong.next.clone()));
         }
 
-        let cost = nn.calculate_costs(test_data);
+        // TODO: Do not calculate_cost another time here
+        let cost = nn.calculate_costs(test_data, pingpong);
         // let test_results = TestResults::new(results, None, avg_cost);
         let test_results = TestResults::new(results, is_correct_fn, cost);
         nn.last_test_results = Some(test_results);
